### DISCUSSION TITLE
Notifications

### DISCUSSION THREAD

**Discussion by bitmvk** [flet-dev/flet#5580] [Extensions]
**Please Describe The Problem To Be Solved**

I want a way to send notifications to the user for example when the app is minimized.

**Comment by ndonkoHenri** [comment]
Some (possibly) useful links:

https://pub.dev/packages/flutter_local_notifications

https://pub.dev/packages/push

**Comment by ZenaMel** [comment]
somewhat related to #547

**Comment by adampiskorski** [comment]
Another package, but this one seems to promise all the same platforms as Flet: https://pub.dev/packages/awesome_notifications, but it seems to require firebase.

I think it may end up being the best to implement notifications yourself in a flutter app that loads flet (https://pub.dev/packages/flet), simply because of the complications caused by each platform, I'm not sure how maintainable and up-to-date an integrated flet solution will be

**Comment by ndonkoHenri** [comment]
Article: https://medium.com/@ndonkohenri/sending-desktop-notifications-in-python-flet-97f0834f993d

Repo: https://github.com/ndonkoHenri/Python-Desktop-Notifications

Hope it helps while waiting for a native solution.

**Comment by arthurbarros** [comment]
> Article: https://medium.com/@ndonkohenri/sending-desktop-notifications-in-python-flet-97f0834f993d
> 
> Repo: https://github.com/ndonkoHenri/Python-Desktop-Notifications
> 
> Hope it helps while waiting for a native solution.

I have come across these links in some Google searches, plyer is a good candidate to solve the problem, what is not clear is how to integrate it with flet. In the test I did, at the moment I execute the notification, the application is closed on the Android emulator.

**Comment by Heoder** [comment]
me too

**Comment by Heoder** [comment]
> Some (possibly) useful links:
> 
> https://pub.dev/packages/flutter_local_notifications
> 
> https://pub.dev/packages/push

How can I add this package to my python project?

**Comment by bitmvk** [comment]
hey, @ndonkoHenri 

now that we have the ability to use flutter packages in flet,

we can make a control using any of the flutter packages to send notification.

Please add this to the roadmap.

**Comment by ndonkoHenri** [comment]
Why close the issue ?

Will add to roadmap.

**Comment by bitmvk** [comment]
> Why close the issue ?

sorry, 
didn't see that i was closing the issue😅

**Comment by chavezbosquez** [comment]
Hello everyone.

I am developing an Android app and I specifically requiere my app to receive push notifications from Firebase via Firebase Cloud Messaging. Does anyone know a workaround to add this funcionality in Flet? Or do I have to wait for a custom component to achieve this?

Thank you all Flet guys for the amazing job so far. ¡Gracias!

**Comment by Heoder** [comment]
> Hello everyone.
> 
> I am developing an Android app and I specifically requiere my app to receive push notifications from Firebase via Firebase Cloud Messaging. Does anyone know a workaround to add this funcionality in Flet? Or do I have to wait for a custom component to achieve this?
> 
> Thank you all Flet guys for the amazing job so far. ¡Gracias!

Oh my god, I'm also really looking forward to having this feature. Currently, flet does not have the push notification feature, while it is a standard feature of a mobile application.

**Comment by MusienkoAnton** [comment]
I'm also looking for push messages for Android in Flet. When we will have them's?

**Comment by lekshmanmj** [comment]
> Some (possibly) useful links:
> 
> https://pub.dev/packages/flutter_local_notifications
> 
> https://pub.dev/packages/push

how about a combination of awesome_notification(for mobile) and flutter_local_notification(for desktop) to flet ?
https://pub.dev/packages/awesome_notifications

**Comment by NHLOCAL** [comment]
I am also very interested in this basic feature.
Has anything progressed in the last month and a half?

**Comment by Muddassir-Farooq-official** [comment]
Kindly add notifications permission in flet build template

**Comment by NHLOCAL** [comment]
Please be more specific: was the feature added in the new version, and which tool should use?

**Comment by Muddassir-Farooq-official** [comment]
Thanks for adding notifications permission

**Comment by Muddassir-Farooq-official** [comment]
Notifications Permission Is Now Enabled In Mobile Devices @ndonkoHenri can u tell me how can we add push notifications using plyer it's work fine on desktop but app crashes on Android can u tell why it's happening ?

**Comment by betopolione** [comment]
Hello everyone. I have a new project to work on and I am looking foward to use flet. However one of the main features of this project is related to remind the user (with push notifications) about some tasks they need to do. I saw this thread and I dont understood if this kind of functionaly is under development or something like this. 

Please, could anyone give me any tip about that.

Thank you guys so much for this amazing framework. So far, loving it. 👍

**Comment by Muddassir-Farooq-official** [comment]
I would recommend you to use flutter package with  flet because this functionality is not available yet

**Comment by betopolione** [comment]
@Muddassir-Farooq-official  tks man

I dont want to look like a slacker, but I would appreciate if you share with me anything about your experience using plyer or other pkg. 

I saw your message above about using plyer and having issues with android. 

Could you please share with me any tip about that? What setup did you use to make it work?

thank you in advance.

**Comment by betopolione** [comment]
hello @Muddassir-Farooq-official 

please, if you have any tip you can give, I really appreciate.

thank you so much.

**Comment by Muddassir-Farooq-official** [comment]
> hello @Muddassir-Farooq-official 
> 
> please, if you have any tip you can give, I really appreciate.
> 
> thank you so much.

Try to implement it with one signal push notifications api and then update me

**Comment by Muddassir-Farooq-official** [comment]
> > hello @Muddassir-Farooq-official 
> > 
> > please, if you have any tip you can give, I really appreciate.
> > 
> > thank you so much.
> 
> Try to implement it with one signal push notifications api and then update me

Any progress? @betopolione

**Comment by betopolione** [comment]
hello @Muddassir-Farooq-official 

not yet, I had to stop to work on it for some days and I am back on it right now. Give me 2 days and should have something to share

thank you

**Comment by betopolione** [comment]
@Muddassir-Farooq-official  

I will ask you a little more time, because I use NixOS and I am fighting lots of issues to have flet working properly

tks

**Comment by betopolione** [comment]
hello @Muddassir-Farooq-official 
sorry for my delay

I manage to make flutter and flet work properly on my NixOS.

Then I created an Flutter test app just to test onesignal notifications and it worked properly.

Now I am in the step where I need to build a flet package to wrap onesignal flutter pkg (https://flet.dev/docs/extend/integrating-existing-flutter-packages).

I am following flet tutoral to create custom pkg, and it says that we need to create a create_control.dart that will return my widget.

however, I am very amateur in flutter, and my questions is that I need to setup OneSignal like this in flutter

```dart
void main() {
  runApp(const MyApp());

  //Remove this method to stop OneSignal Debugging
  OneSignal.Debug.setLogLevel(OSLogLevel.verbose);

  OneSignal.initialize("123-123-123-132-123");

  // The promptForPushNotificationsWithUserResponse function will show the iOS or Android push notification prompt. We recommend removing the following code and instead using an In-App Message to prompt for notification permission
  OneSignal.Notifications.requestPermission(true);

}
```

And I am triyng to understand how this is translanted to flet package, since this is about a config done when starting up the app, instead of a simple Widget that is drawed on screen.

Could you please give em any direction?

thank you so much

**Comment by FeodorFitsner** [comment]
Is https://pub.dev/packages/onesignal_flutter what we need to support proper notifications in Flet?

**Comment by Muddassir-Farooq-official** [comment]
> Is https://pub.dev/packages/onesignal_flutter what we need to support proper notifications in Flet?

You can do it using awesome notifications flutter package or you can make python cross platform library player compatible to flet for local notification while for push notifications you can go with Firebase or one signal but highly recommended is Firebase if possible

**Comment by betopolione** [comment]
Hello @FeodorFitsner and @Muddassir-Farooq-official ,

thank you for all your support,
 
I am trying to create an extension using **onesignal_flutter** package .

This flutter pkg needs to be initialized and after doing some researches I think I found the answer. As I understood I just need to add the inializations code in the **create_control.dart** file

```dart
void ensureInitialized() {
  // nothing to initialize
}
```

in order to undertand all the proccess of how to integrate existing flutter packages in the Flet framework, I am trying to follow this simple tutorial
https://flet.dev/docs/extend/integrating-existing-flutter-packages


I am getting an weird error when running the build command:
```
Because src depends on flutter from sdk which doesn't exist (the Flutter SDK is not available), version solving failed.

Flutter users should use `flutter pub` instead of `dart pub`.
```

Here is the full output, I am runnning it from my flet project src folder

```
[beto@nixos:~/flet_tests/src]$ flet build --verbose --template https://github.com/flet-dev/flet-build-template --template-ref 0.24.1 linux

[10:20:27] Flutter bootstrap directory: /tmp/nix-shell.wnPoAi/flet_flutter_build_WU8pvPy0Mz                                                  
           Additional Flutter dependencies: {'flet_spinkit': {'path': '/home/beto/flutter_tests/DART_PKG/flet_spinkit'}}                   
[10:20:29] Created Flutter bootstrap project from https://github.com/flet-dev/flet-build-template with ref 0.24.1 ✅                         
           Copying /home/beto/flet_tests/src/assets/icon.png to /tmp/nix-shell.wnPoAi/flet_flutter_build_WU8pvPy0Mz/images                 
           Customized app icons and splash images ✅                                                                                         
           Run subprocess: ['/nix/store/rjz395nsk0zb67gaxkjc9qpx0khvl9d3-flutter-wrapped-3.24.4-sdk-links/bin/dart', 'run',                  
           'flutter_launcher_icons']                                                                                                         
( ●    ) Generating app icons (1.7s)
(  ●   ) Generating app icons ⏳... Because src depends on flutter from sdk which doesn't exist (the Flutter SDK is not available), version solving failed.

Flutter users should use `flutter pub` instead of `dart pub`.

[10:20:32] Deleting Flutter bootstrap directory /tmp/nix-shell.wnPoAi/flet_flutter_build_WU8pvPy0Mz                                          
           Error building Flet app - see the log of failed command above.                                                                    
           Run subprocess: ['/nix/store/rjz395nsk0zb67gaxkjc9qpx0khvl9d3-flutter-wrapped-3.24.4-sdk-links/bin/flutter', '--version']         
           Run subprocess: ['/nix/store/rjz395nsk0zb67gaxkjc9qpx0khvl9d3-flutter-wrapped-3.24.4-sdk-links/bin/flutter', 'doctor']            
[10:20:37] Doctor summary (to see all details, run flutter doctor -v):                                                                       
           [✓] Flutter (Channel stable, 3.24.4, on NixOS 24.11 (Vicuna) 6.6.63, locale en_US.UTF-8)                                          
           [✓] Android toolchain - develop for Android devices (Android SDK version 32.0.0)                                                  
           [✗] Chrome - develop for the web (Cannot find Chrome executable at google-chrome)                                                 
               ! Cannot find Chrome. Try setting CHROME_EXECUTABLE to a Chrome executable.                                                   
           [✓] Linux toolchain - develop for Linux desktop                                                                                   
           [!] Android Studio (not installed)                                                                                                
           [✓] Connected device (1 available)                                                                                                
           [✓] Network resources                                                                                                             
                                                                                                                                             
           ! Doctor found issues in 2 categories.   
```

Please, let me know any other detail you guys need and I will provide.

thank you so much

**Comment by brunobrown** [comment]
@betopolione 

Hi bro,
Maybe this video in the link below or the source code in the repository can help you. Feel free to use it or contribute if you'd like.


- https://discord.com/channels/981374556059086931/1247163952186654790/1319883478451490898

- https://github.com/brunobrown/onesignal-flet

**Comment by betopolione** [comment]
hello @brunobrown , thank you very much, man, for providing this sample and this video demonstration, amazing!!!

Do you know if is it possible to run `flet run` (afer flet build) and be able to see the flutter component (package) in development time? Imagine you are customizing the component (a visual component), will you always need to build the apk just for checking some customization?

I saw that when we send push notification with onesignal, the android only shows a bell at the top of the screen, but does not display that component similar to a dialog with a preview of the notification (that disappears automatically after +-5 seconds). Is that another component?

PS: Are you from Brasil? (vi q seu OBS estava em pt-br.)

thank you very much, your code sample is helping me a lot understand more about flet integration with flutter packages

**Comment by brunobrown** [comment]
1 - Unfortunately it is not possible to see this at run time.
2 - Regarding OneSignal and the dialog box, I believe this is more of a configuration directly in the OneSignal platform dashboard. I haven't thoroughly explored all the possibilities yet.
3 - Yes, I’m from Brazil. By the way, I apologize for my English as I’m still learning and often use Google Translate and ChatGPT to help with it.

PS: I’m forwarding the first question to @FeodorFitsner and @ndonkoHenri: "Do you know if it’s possible to run flet run (after building Flet) and view the Flutter component (package) during development time?". That’s a great question!

**Comment by betopolione** [comment]
Hey Man, I am from Brazil too, and my english is not so good, but I didnt notice any problems with yours, on the contrary ✌️😃

and, again, thank you very much for your help, your code sample of integration gave me lots of insights

would be awesome in the future be able to see changes in Flutter component (package) during development time, without need `flet build`, but this is just a little detail, this framework is amazing. The magic done by these guys is incredible

🚀

**Comment by Muddassir-Farooq-official** [comment]
![image](https://github.com/user-attachments/assets/297d111e-75cf-4094-b5b2-b4048c318dd1)

irs says unknown control flet

**Comment by brunobrown** [comment]
@Muddassir-Farooq-official 

Hi bro,
Did you configure the path for the Flutter package dependency in your pyproject.toml file?

Example:

```toml

[tool.poetry]  
name = "example-flet-app"  
version = "0.1.0"  
description = ""  
authors = ["brunobrown <example@test.com>"]  
readme = "README.md"  

[tool.flet.flutter.dependencies]  # <--- here  
flet_onesignal = "^0.0.1"  # <--- package here  

```

More details about this configuration can be found at:

https://flet.dev/blog/pyproject-toml-support-for-flet-build-command/#flutter-settings

**Comment by Muddassir-Farooq-official** [comment]
Can I test it through flet run main.py without building apk?

**Comment by brunobrown** [comment]
Unfortunately it is not possible. You need to perform the build.

**Comment by Muddassir-Farooq-official** [comment]
> Unfortunately it is not possible. You need to perform the build.

https://discord.com/channels/981374556059086931


This link is not working... 

![Screenshot_20241224-221253.jpg](https://github.com/user-attachments/assets/7af8fdac-47c7-4be5-b82f-e672c1530249)

**Comment by brunobrown** [comment]
@Muddassir-Farooq-official

Did you mean the link to the video? If so, the full link is this:

https://discord.com/channels/981374556059086931/1247163952186654790/1319883478451490898

**Comment by Muddassir-Farooq-official** [comment]
> @Muddassir-Farooq-official
> 
> Did you mean the link to the video? If so, the full link is this:
> 
> https://discord.com/channels/981374556059086931/1247163952186654790/1319883478451490898 

Yes but still this link is not working

**Comment by brunobrown** [comment]
Flet OneSignal - Flutter OneSignal package integration.

- https://github.com/brunobrown/onesignal-flet

- https://github.com/user-attachments/assets/687dae56-bd80-4c2f-a613-35ab7d2ddc56

**Comment by Muddassir-Farooq-official** [comment]
> Flet OneSignal - Flutter OneSignal package integration.
> 
> - https://github.com/brunobrown/onesignal-flet
> 
> - https://github.com/user-attachments/assets/687dae56-bd80-4c2f-a613-35ab7d2ddc56

To send *specific notifications to a specific user's device* using OneSignal in your Flutter app, you'll need to identify and target that specific device (or user). OneSignal provides several ways to achieve this, such as *tags*, *player IDs*, or *segments*.

Here’s a step-by-step guide to send notifications to a *specific device* using OneSignal:

1. *Identify the User Device*
To send notifications to a specific user, you need to capture and store a unique identifier for that user’s device. OneSignal provides a *Player ID* for each device, which is unique to that device and is required to target notifications to a specific device.

*Getting the OneSignal Player ID in Flutter*:
Once a device is registered with OneSignal, it gets a *Player ID*. You can capture this ID and store it in your backend or database for later use.

Here’s how you can get the *Player ID* in your Flutter app:

```dart
import 'package:onesignal_flutter/onesignal_flutter.dart';

void main() {
  runApp(MyApp());

  // Initialize OneSignal
  OneSignal.shared.setAppId("YOUR_ONESIGNAL_APP_ID");

  // Get the Player ID (device identifier)
  OneSignal.shared.getDeviceState().then((deviceState) {
    String playerId = deviceState?.userId ?? "No Player ID";
    print("Player ID: $playerId");

**Comment by Muddassir-Farooq-official** [comment]
How can we get user device player I'd using this flet-onesignal

**Comment by brunobrown** [comment]
@Muddassir-Farooq-official

Hi bro,

What I did was just the beginning of a potential integration for the package. Things have been quite hectic for me lately with other tasks and activities, and I haven’t had the time to dive into the Flutter package to add new features to flet-onesignal.

Also, my knowledge of Flutter is minimal or almost nonexistent, so I need to study and understand the OneSignal Flutter logic to implement it in flet-onesignal. If you'd like to help implement new features, I’d be grateful and would be more than happy to add you as a collaborator on the project. 

What do you think?

My invitation also extends to others involved in this issue who may want to contribute.

**Comment by Muddassir-Farooq-official** [comment]
> > Flet OneSignal - Flutter OneSignal package integration.
> > 
> > * https://github.com/brunobrown/onesignal-flet
> > * 2024-12-21_01-13-47.1.mp4
> 
> To send _specific notifications to a specific user's device_ using OneSignal in your Flutter app, you'll need to identify and target that specific device (or user). OneSignal provides several ways to achieve this, such as _tags_, _player IDs_, or _segments_.
> 
> Here’s a step-by-step guide to send notifications to a _specific device_ using OneSignal:
> 
> 1. _Identify the User Device_
>    To send notifications to a specific user, you need to capture and store a unique identifier for that user’s device. OneSignal provides a _Player ID_ for each device, which is unique to that device and is required to target notifications to a specific device.
> 
> _Getting the OneSignal Player ID in Flutter_: Once a device is registered with OneSignal, it gets a _Player ID_. You can capture this ID and store it in your backend or database for later use.
> 
> Here’s how you can get the _Player ID_ in your Flutter app:
> 
> ```dart
> import 'package:onesignal_flutter/onesignal_flutter.dart';
> 
> void main() {
>   runApp(MyApp());
> 
>   // Initialize OneSignal
>   OneSignal.shared.setAppId("YOUR_ONESIGNAL_APP_ID");
> 
>   // Get the Player ID (device identifier)
>   OneSignal.shared.getDeviceState().then((deviceState) {
>     String playerId = deviceState?.userId ?? "No Player ID";
>     print("Player ID: $playerId");
> ```

its stucks during build apk for android at


[14:17:54] Generated splash screens ✅
           Run subprocess: ['C:\\src\\flutter_windows_3.19.6-stable\\flutter\\bin\\dart.BAT', 'run', 'serious_python:main', 'package', 'G:\\Projects\\Bistro New', '--platform', 'Android', '--requirements',  
           'flet>=0.25.2,onesignal_flet@git+https://github.com/brunobrown/onesignal-flet.git@main', '--exclude', 'build', '--cleanup', '--verbose']
(  ●   ) Packaging Python app ⏳...

**Comment by brunobrown** [comment]
Hi bro. 

I use Flutter v3.24.2 and Poetry for managing and configuring my projects. All configuration is done in the ```pyproject.toml``` file. Then, I simply run ```flet build apk```.

Below is an example of a pyproject.toml file configuration in case you also use Poetry:

```toml

[tool.poetry]  
name = "example-flet-app"  
version = "0.1.0"  
description = ""  
authors = ["brunobrown <example@test.com>"]  
readme = "README.md" 

[tool.poetry.dependencies]
python = "^3.12"
flet = "^0.25.2"
onesignal_flet = {git = "https://github.com/brunobrown/onesignal-flet.git", rev = "main"}

[tool.flet]
org = "com.mycompany" # --org
product = "Product name" # --product
company = "My Company" # --company
copyright = "Copyright (C) 2024 by MyCompany" # --copyright
build_number = 1 # --build-number
app.module = "main" # --module-name
compile.cleanup = false # --cleanup-on-compile

[tool.flet.flutter.dependencies]  # <--- here  
flet_onesignal = "^0.0.1"  # <--- Flutter package here

[build-system]
requires = ["poetry-core"]
build-backend = "poetry.core.masonry.api"

```

Even if you don't use Poetry, you can just create a ```pyproject.toml``` file in the root of your project and follow the various configuration examples shared [here](https://flet.dev/blog/pyproject-toml-support-for-flet-build-command) by Feodor:

**Comment by Muddassir-Farooq-official** [comment]
> @betopolione 
> 
> Hi bro,
> Maybe this video in the link below or the source code in the repository can help you. Feel free to use it or contribute if you'd like.
> 
> 
> - https://discord.com/channels/981374556059086931/1247163952186654790/1319883478451490898
> 
> - https://github.com/brunobrown/onesignal-flet
> 
> 

@brunobrown can you merge it in flet as a pull request

**Comment by Muddassir-Farooq-official** [comment]
@FeodorFitsner as flet now supporting pyjnius when will flet support plyer?

**Comment by Ayberk-Irmak** [comment]
Hey everyone! 👋

I finally found a solution for sending notifications in Flet mobile applications using Pyjnius!
Since I know that many of you have been struggling with this issue, I decided to make a video tutorial explaining the entire process.
Check it out, and let me know what you think!

🔗 Video Link: [How to Send Notifications in Flet using Pyjnius](https://www.youtube.com/watch?v=kXWKaylIhCg)

Hope it helps! 😊

**Comment by Muddassir-Farooq-official** [comment]
Here is solution for local push notifications in flet

https://github.com/Bbalduzz/flet_notifications

**Comment by alejandropoos** [comment]
I've noticed that the solution for @Ayberk-Irmak and @Muddassir-Farooq-official  doesn't allow using ft.Container, ft.Buttons, or any other non-string control as notification content. How could you put, for example, a container with buttons inside the notification content?

**Comment by Sushmargowda27** [comment]
Hey everyone! 👋
I’ve successfully implemented push notifications on Android in my Flet mobile app using Pyjnius.
Now, I’m trying to get push notifications working on iOS, but I’m stuck. 😕
Has anyone managed to enable iOS push notifications in a Flet app?
What’s the right approach or library for iOS?
Do I need to configure anything specific in Xcode or Apple Developer settings (like provisioning profiles, certificates, etc.)?
Do I need to use Swift/Objective-C or is there a Python-based solution for iOS like Pyjnius for Android?
Any help or guidance is greatly appreciated — thanks in advance! 🙏
